### ISSUE TITLE
FreeCAD Path needs a new Postprocessor architecture.

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#6117]
Issue imported from https://tracker.freecad.org/view.php?id=4468

* **Reporter:** sliptonic
* **Date submitted:** 10/18/2020
* **FreeCAD version:** 
* **Category:** Feature
* **Status:** new
* **Tags:** 

# Original report text

The current post-processor architecture requires individual .py files for each post.  This is flexible in allowing users to copy an existing post and modify it but results in lots of duplicated code and redundant work.
Changes and improvements made in one post don't become part of the other posts so bugs and incompatibilities persist.

# Additional information

A better architecture would allow code reuse while preserving the users ability to customize the post to their needs.

# Other bug information

* **Priority:** normal
* **Severity:** feature
* **Category:** Feature
* **Updated:** 2/6/2021

**Comment by LarryWoestman** [comment]
I am currently working on refactoring the postprocessor code and the postprocessors.

I think it makes sense to assign this issue to me.

**Comment by luzpaz** [comment]
@sliptonic :point_up:

**Comment by luzpaz** [comment]
@sliptonic JFYI, this ticket is still open yet is labeled 'Done' in your CAM stuff list.

**Comment by sliptonic** [comment]
Yeah, I left it open because the refactoring only touched the first part of post-processing.  The next step will be to start refactoring the individual posts to re-use code and to be easier to customize.

**Comment by LarryWoestman** [comment]
Progress is being made on refactoring the individual posts, with more work to come.

**Comment by luzpaz** [comment]
@LarryWoestman what's your verdict? can we close this now ?

**Comment by LarryWoestman** [comment]
Not yet.  I would consider the process over half done, but still more work to go.  The main tasks left relate to adding back in enough customizability to make up for having lost the ability to change the shared code.  There are also over 10 more postprocessors that I would like to port to the "refactored" code base.  I suspect that I will learn about some additional customizability requirements from each one :-)
